lib/src/front_end/piece_writer.dart

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import '../back_end/solver.dart';
+import '../dart_formatter.dart';
+import '../debug.dart' as debug;
+import '../piece/piece.dart';
+import '../source_code.dart';
+
+/// Incrementally builds [Piece]s while visiting AST nodes.
```

Comment:
Why would a "Writer" be the one that "Builds" pieces? I would expect the "Factory" to "Build" them...

Reply:
Oh, God, I went back and forth on how to name these damn classes for *ages*.

I don't love any of the names, but my general thinking was to use "Writer" for classes that incrementally build something up. They are internally mutable and "own" the thing being built (which they generally yield at the end). So `PieceWriter` is used to incrementally build the piece tree. (I suppose I could have called it `PieceTreeWriter`, but that seems unnecessary.) And `CodeWriter` incrementally builds the final output code.

I considered "Builder" for these, but that felt even more ambiguous with "Factory".

I used "Factory" to refer to functionality that produces new objects but doesn't have any interesting state or mutability itself. It's a little odd here because usually factory methods *return* the thing they make. I tried to do that here, but it doesn't work well because the `Piece` tree node boundaries often don't align well with the AST node boundaries. That's why `PieceWriter` has the `push()` and `pop()` API.

(I could have forced the piece tree to align with the AST node boundaries so that every AST node is a piece. But that would lead to a much larger piece tree with many places where no split can occur between a piece boundary anyway. I worry that would bleed perf for little benefit. My goal is to make the piece tree as coarse-grained as possible because that gives the back end, which is where all the perf challenges are, less to chew on.)

I had `PieceMaker` for a while too, which I loved for the homophonic pun, but otherwise didn't seem to clarify. Naming is hard.

The other option would be to lump it all together. There's nothing preventing everything from being right inside AstNodeVisitor but then it becomes a giant monster of a class.